include/internal/rcu.h

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OPENSSL_RCU_H
+# define OPENSSL_RCU_H
+# pragma once
+
+typedef void (*rcu_cb_fn)(void *data);
+
+typedef void CRYPTO_RCU_LOCK;
```

Comment:
This is slightly odd. There is no reason that this shouldn't just be:

````
typedef struct rcu_lock_st CRYPTO_RCU_LOCK;
````
The structure definition can remain where it is now, and the struct is just opaque to all other callers. Then you don't need to do all the casting to `struct rcu_lock_st` that you currently do.

Reply:
It is indeed how CRYPTO_RWLOCK is defined. IMO that is a mistake and because of it we don't get proper type checking with it. Which has caused problems before.

Reply:
> It is indeed how CRYPTO_RWLOCK is defined. IMO that is a mistake and because of it we don't get proper type checking with it. Which has caused problems before.

Exactly. CRYPTO_RWLOCK is the exception to the rule of how we do it virtually everywhere else.

Reply:
modified.